### PR TITLE
Link standalone discovery image information during build time

### PR DESCRIPTION
**Pull request by anujc25**
### What this PR does / why we need it
- This PR implements build-time linking of standalone discovery image
- Allows user to override configuration with runtime environment variables
- Also adds base for Kubernetes Discovery Support
- Unit tests

Build time variables introduced:
- `TKG_DEFAULT_IMAGE_REPOSITORY` (already exists before this change) 
- `DEFAULT_STANDALONE_DISCOVERY_IMAGE_PATH` 
- `DEFAULT_STANDALONE_DISCOVERY_IMAGE_TAG`

Users can override standalone discovery at runtime with below environment variables:
- `TKG_CUSTOM_IMAGE_REPOSITORY`
- `TKG_DEFAULT_STANDALONE_DISCOVERY_IMAGE_PATH`
- `TKG_DEFAULT_STANDALONE_DISCOVERY_IMAGE_TAG`

### Which issue(s) this PR fixes

Related to #74 

### Describe testing done for PR

- Manually tested the CLI by providing build-time configuration and testing it to determine the default standalone discovery is getting used correctly

### Release note

```release-note
NONE
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: pkg/v1/cli/discovery/kubernetes.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+// Name of the repository.
+func (k *KubernetesDiscovery) Name() string {
+	return k.name
+}
+
+// Manifest returns the manifest for a local repository.
+func (k *KubernetesDiscovery) Manifest() ([]plugin.Discovered, error) {
+	plugins := make([]plugin.Discovered, 0)
+	// TODO: Implement Kubernetes Discovery
```
Comment (anujc25) [inline pkg/v1/cli/discovery/kubernetes.go:58]: This will be implemented as part of separate PR.

**Review by saji-pivotal — APPROVED** [review]
Looks good to me. But I would definitely like another person to take a look at this as well

File: pkg/v1/cli/pluginmanager/manager.go [hunk 2]
```diff
@@ -113,15 +113,30 @@ func DiscoverServerPlugins(serverName string) (plugins []plugin.Discovered, err
 		return
```
Comment (imikushin) [inline pkg/v1/cli/pluginmanager/manager.go:113]: The returned error here is nil, and I guess that's not intended. This is a consequence of using return vars. 

Generally, you don't need named return vars. Sure, it makes your `return` statements shorter, but that's also a source of confusion for the reader (future you :P). Instead of immediately seeing what's returned you have to traverse back to see what was last assigned to the return var.

Please make your `return` statements explicit. 

And, the tests did *not* catch it, so add a test :)

File: pkg/v1/config/clientconfig.go [hunk 3]
```diff
@@ -143,6 +146,75 @@ func NewClientConfig() (*configv1alpha1.ClientConfig, error) {
...
+		Local: &configv1alpha1.LocalDiscovery{
+			Name: DefaultStandaloneDiscoveryName,
+			Path: DefaultStandaloneDiscoveryLocalPath,
+		},
+	}
+
+	// Prepend default discovery to available discovery sources
+	c.ClientOptions.CLI.DiscoverySources = append([]configv1alpha1.PluginDiscovery{defaultDiscovery}, c.ClientOptions.CLI.DiscoverySources...)
+	return true
+}
```
Comment (imikushin) [inline pkg/v1/config/clientconfig.go:192]: This function doesn't seem to be covered by unit tests.

File: pkg/v1/config/defaults.go [hunk 4]
```diff
@@ -37,3 +53,24 @@ var DefaultRepositories []configv1alpha1.PluginRepository = []configv1alpha1.Plu
...
+	// Run-time overrides of the configuration
+	if customImageRepo := os.Getenv(constants.ConfigVariableCustomImageRepository); customImageRepo != "" {
+		defaultStandaloneDiscoveryRepository = customImageRepo
+	}
+	if imagePath := os.Getenv(constants.ConfigVariableDefaultStandaloneDiscoveryImagePath); imagePath != "" {
+		defaultStandaloneDiscoveryImagePath = imagePath
+	}
+	if imageTag := os.Getenv(constants.ConfigVariableDefaultStandaloneDiscoveryImageTag); imageTag != "" {
+		defaultStandaloneDiscoveryImageTag = imageTag
+	}
```
Comment (imikushin) [inline pkg/v1/config/defaults.go:73]: All these `if` conditions seem to eval to false when running unit tests, so they're only "half" covered.

**Review by imikushin — COMMENTED** [review]
A few comments. Sorry for being picky 🙂

File: pkg/v1/cli/pluginmanager/manager.go (hunk 2 above)
Comment (anujc25) [reply pkg/v1/cli/pluginmanager/manager.go:113]: This is intended. If the current server provided is incorrect, I believe we should throw a warning indicating the message instead of throwing an error. (added a warning message)

This will allow users to run `tanzu plugin list`, `tanzu plugin sync` etc commands to sync plugins with a warning message which indicates `Error while discovering contexted plugins`.  I feel we should not block users from running other plugin commands if something is misconfigured (especially someone mistakenly misconfigures tanzu config file)

File: pkg/v1/cli/discovery/interface.go [hunk 5]
```diff
@@ -41,5 +41,8 @@ func CreateDiscoveryFromV1alpha1(pd v1alpha1.PluginDiscovery) (Discovery, error)
 	if pd.Local != nil {
 		return NewLocalDiscovery(pd.Local.Name, pd.Local.Path), nil
 	}
+	if pd.Kubernetes != nil {
+		return NewKubernetesDiscovery(pd.Kubernetes.Name, pd.Kubernetes.Path, pd.Kubernetes.Context), nil
+	}
```
[the remark this answers is not in the document]
Comment (anujc25) [reply pkg/v1/cli/discovery/interface.go:46]: updated and added unit tests

File: pkg/v1/config/clientconfig.go (hunk 3 above)
Comment (anujc25) [reply pkg/v1/config/clientconfig.go:191]: added unit test

File: pkg/v1/config/defaults.go (hunk 4 above)
Comment (anujc25) [reply pkg/v1/config/defaults.go:73]: added unit tests

File: pkg/v1/config/clientconfig.go (hunk 3 above)
Comment (imikushin) [reply pkg/v1/config/clientconfig.go:191]: Thanks for adding tests! The `for` loop body is still not covered, and it contains non-trivial logic. Perhaps, it even makes sense to extract it as a separate function.

File: pkg/v1/config/clientconfig.go [hunk 6]
```diff
@@ -507,3 +580,30 @@ func IsFeatureActivated(feature string) bool {
...
+	if server.Type == configv1alpha1.ManagementClusterServerType {
+		defaultClusterK8sDiscovery := configv1alpha1.PluginDiscovery{
+			Kubernetes: &configv1alpha1.KubernetesDiscovery{
+				Name:    fmt.Sprintf("default-%s", serverName),
+				Path:    server.ManagementClusterOpts.Path,
+				Context: server.ManagementClusterOpts.Context,
+			},
+		}
+		discoverySources = append(discoverySources, defaultClusterK8sDiscovery)
+	}
```
Comment (imikushin) [inline pkg/v1/config/clientconfig.go:607]: The `if` body doesn't seem to be covered by tests.

This method does get called (indirectly) from manager_test.go, but it doesn't have "its own" unit tests.

**Review by imikushin — COMMENTED** [review]
A couple more comments

File: pkg/v1/cli/pluginmanager/manager.go [hunk 7]
```diff
@@ -102,26 +102,24 @@ func DiscoverStandalonePlugins() (plugins []plugin.Discovered, err error) {
 }
 
 // DiscoverServerPlugins returns the available plugins associated with the given server
-func DiscoverServerPlugins(serverName string) (plugins []plugin.Discovered, err error) {
-	plugins = []plugin.Discovered{}
+func DiscoverServerPlugins(serverName string) ([]plugin.Discovered, error) {
+	plugins := []plugin.Discovered{}
```
[the remark this answers is not in the document]
Comment (anujc25) [reply pkg/v1/cli/pluginmanager/manager.go:106]: I would prefer not to return `nil` and return an empty list.

File: pkg/v1/config/clientconfig.go (hunk 6 above)
Comment (anujc25) [reply pkg/v1/config/clientconfig.go:607]: added unit test for this function which includes `if` body.

File: pkg/v1/cli/pluginmanager/manager.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (imikushin) [reply pkg/v1/cli/pluginmanager/manager.go:106]: If `plugins == nil` the following code works:
`plugins = append(plugins, &plugin.Discovered{})`

**Review by imikushin — APPROVED** [review]
Looks good!
